src/axom/sidre/core/Group.cpp

```diff
@@ -104,6 +104,136 @@ std::string Group::getPath() const
   return thePath;
 }
 
+/*
+ *************************************************************************
+ *
+ * Return information about data associated with Group subtree with this 
+ * Group at root of tree (default 'recursive' is true), or for this Group 
+ * only ('recursive' is false) in fields of given Conduit Node.
+ *
+ *************************************************************************
+ */
+void Group::getDataInfo(Node& n, bool recursive) const
+{
+  //
+  // Initialize Node fields
+  //
+  IndexType num_groups = 0;
+  IndexType num_views = 0;
+  IndexType num_views_empty = 0;
+  IndexType num_views_buffer = 0;
+  IndexType num_views_external = 0;
+  IndexType num_views_scalar = 0;
+  IndexType num_views_string = 0;
+  IndexType num_bytes_allocated = 0;
+  IndexType num_bytes_external = 0;
+
+  n["num_groups"] = num_groups;
+  n["num_views"] = num_views;
+  n["num_views_empty"] = num_views_empty;
+  n["num_views_buffer"] = num_views_buffer;
+  n["num_views_external"] = num_views_external;
+  n["num_views_scalar"] = num_views_scalar;
+  n["num_views_string"] = num_views_string;
+  n["num_bytes_allocated"] = num_bytes_allocated;
+  n["num_bytes_external"] = num_bytes_external;
+
+  getDataInfoHelper(n, recursive);
+}
+
+/*
+ *************************************************************************
+ *
+ * Private helper method to support getDataInfo() method.
+ *
+ *************************************************************************
+ */
+void Group::getDataInfoHelper(Node& n, bool recursive) const
+{
+  //
+  // Grab Node entries for updating data info for this Group
+  //
+  IndexType num_groups = n["num_groups"].value();
+  IndexType num_views = n["num_views"].value();
+  IndexType num_views_empty = n["num_views_empty"].value();
+  IndexType num_views_buffer = n["num_views_buffer"].value();
+  IndexType num_views_external = n["num_views_external"].value();
+  IndexType num_views_scalar = n["num_views_scalar"].value();
+  IndexType num_views_string = n["num_views_string"].value();
+  IndexType num_bytes_allocated = n["num_bytes_allocated"].value();
+  IndexType num_bytes_external = n["num_bytes_external"].value();
+
+  num_groups += 1;  // count this group
+
+  //
+  // Gather info from Views owned by this Group
+  //
+  IndexType vidx = getFirstValidViewIndex();
+  while(indexIsValid(vidx))
+  {
+    const View* view = getView(vidx);
+
+    num_views += 1;
+
+    if(view->isExternal())
+    {
+      num_views_external += 1;
+      num_bytes_external += view->getTotalBytes();
+    }
+    else if(view->isScalar())
+    {
+      num_views_scalar += 1;
+      num_bytes_allocated += view->getTotalBytes();
+    }
+    else if(view->isString())
+    {
+      num_views_string += 1;
+      num_bytes_allocated += view->getTotalBytes();
+    }
+    else if(view->hasBuffer())
+    {
+      num_views_buffer += 1;
+      if(view->isAllocated())
+      {
+        num_bytes_allocated += view->getTotalBytes();
+      }
```

Comment:
If multiple Views use the same Buffer, I think this will redundantly accumulate the byte count for the same data.  Though perhaps this is desired.  The same thing applies to multiple Views using the same external data.

Reply:
I thought about that and was waiting for someone to bring it up. It will only double count if two (or more) views represent data that *overlaps in a single buffer*.  To me, this seems like a very rare case. To correctly account for that case gets complicated quickly. Need to track buffer ids so that I know when a buffer is assoc with > 1 view. Then I need to look at the views attached to the buffer, figure out which ones are in the group subtree I am working with and whether I've seen one of those before. So I would need to know which groups have been processed already, etc. That seems like a lot of work for a corner case. Think about the extra complications for offsets and non-unit strides.

More simply, I could just track buffer ids and accumulate all the bytes in a buffer once if any view I see in the traversal is associated with it. That's much easier, but may also not be what a user expects since it could report more bytes than they think they are looking.

Similarly for external views. I would have to track memory addresses, offsets, strides, etc.

What do folks think?

Reply:
I agree that it's not worth the trouble to analyze the schemas or compare addresses.  I think the label "num_bytes_allocated" can be confusing if it's not truly a measure of the size of memory allocated.  Maybe something like "num_bytes_visible"?

Reply:
How about 'num_bytes_assoc_with_views'?